src/main/java/com/amazonaws/encryptionsdk/model/EncryptionMaterials.java

```diff
@@ -70,20 +71,28 @@ public CryptoAlgorithm getAlgorithm() {
      * The KeyBlobs to serialize (in cleartext) into the encrypted message.
      */
     public List<KeyBlob> getEncryptedDataKeys() {
-        return unmodifiableList(encryptedDataKeys);
+        return encryptedDataKeys;
     }
 
     /**
-     * Add an encrypted data key to the list of encrypted data keys.
+     * Creates a new {@code EncryptionMaterials} instance based on this instance with the addition of the
+     * provided encrypted data key and keyring trace entry.
      *
      * @param encryptedDataKey  The encrypted data key to add.
      * @param keyringTraceEntry The keyring trace entry recording this action.
+     * @return The new {@code EncryptionMaterials} instance.
      */
-    public void addEncryptedDataKey(KeyBlob encryptedDataKey, KeyringTraceEntry keyringTraceEntry) {
+    public EncryptionMaterials withEncryptedDataKey(KeyBlob encryptedDataKey, KeyringTraceEntry keyringTraceEntry) {
         requireNonNull(encryptedDataKey, "encryptedDataKey is required");
         requireNonNull(keyringTraceEntry, "keyringTraceEntry is required");
+
+        final List<KeyBlob> encryptedDataKeys = new ArrayList<>(getEncryptedDataKeys());
         encryptedDataKeys.add(encryptedDataKey);
-        keyringTrace.add(keyringTraceEntry);
+
+        return toBuilder()
+                .setEncryptedDataKeys(encryptedDataKeys)
+                .setKeyringTrace(keyringTrace.with(keyringTraceEntry))
```

Comment:
Nit: shouldn't these be "with" methods?

(Note I'm not 100% sure if the builder existed already, if so then don't change this)

Reply:
The builder already existed